### PR TITLE
Use string.IsNullOrEmpty and ArgumentException.ThrowIfNullOrEmpty in many more places

### PR DESCRIPTION
**Pull request by stephentoub**
We have an analyzer for ThrowIfNullOrEmpty; I'll separately look into why it wasn't firing.

We should consider an analyzer for string.IsNullOrEmpty.  There's no perf benefit today, other than smaller IL, but the resulting code is easier to read IMHO, and there can be a benefit in the few cases where a complicated expression is duplicated and the JIT can't CSE the whole thing.

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/Resources/Strings.resx [hunk 1]
```diff
@@ -3869,6 +3866,9 @@
   <data name="Argument_EmptyString" xml:space="preserve">
     <value>The value cannot be an empty string.</value>
   </data>
+  <data name="Argument_EmptyOrWhiteSpaceString" xml:space="preserve">
```
Comment (MihaZupan) [inline src/libraries/System.Private.CoreLib/src/Resources/Strings.resx:3869]: Is this used anywhere?

File: src/libraries/System.Private.CoreLib/src/Resources/Strings.resx [hunk 2]
```diff
@@ -1663,9 +1663,6 @@
   <data name="ArgumentNull_AssemblyName" xml:space="preserve">
     <value>AssemblyName cannot be null.</value>
   </data>
-  <data name="ArgumentNull_AssemblyNameName" xml:space="preserve">
```
Comment (MihaZupan) [inline src/libraries/System.Private.CoreLib/src/Resources/Strings.resx:1666]: Is this being referenced from the native side?
https://github.com/dotnet/runtime/blob/7bd46666716a0c7311b0e28594238fe4fa2d3b60/src/coreclr/vm/assembly.cpp#L397

**Review by MihaZupan — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/Resources/Strings.resx (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/Resources/Strings.resx:3869]: Nope, it's a holdover from another branch where I was adding ThrowIfNullOrWhiteSpace.

File: src/libraries/System.Private.CoreLib/src/Resources/Strings.resx (hunk 2 above)
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/Resources/Strings.resx:1666]: Yup

**Comment by stephentoub** [comment]
> Unfortunately, forward substitution still does not handle `IsNullOrEmpty()` vs. `? true : false`, as latter is better in some cases. i.e. in https://godbolt.org/z/rc4Kx56aq M2 is worse than M3 and M1 has the best code.

Is this just a general comment or is this relevant to the changes made in this PR?  This PR didn't remove any uses of `? true : false`, and the differences between `s is null || s.Length == 0` vs `string.IsNullOrEmpty()` are purely about the number of returns:
https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+BiAdgVwDZ4CGweMABACYCWAzsaQLABQAAgMxksCMSnATGQDCZAN7MyEzh269gECHjIBZLgApuABjI0AlGQC8APm1laZXATJo02gHQAZGFgDmAFwAWB/WQ0BucZLsnDxkcgrKfOpcWroGxpq2AJI0AHL4eADyUACiALYADq4Anqq6/kwAvkA===
and even that difference evaporates if you do anything other than directly return the result of the operation, which also doesn't happen in this PR.

**Comment by MichalPetryka** [comment]
> are purely about the number of returns as used by the C# compiler

In [this](https://godbolt.org/z/5nfdcqh9E) case [the number of returns matches](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+BiAdgVwDZ4CGweMABACYCWAzsaQLABQzAAgMxmsCMSZOWOgDNyrAExkAKjBoAXZgG9mZFV048+wCBDxkAggAoeABjI0AlMtVKmqu1wDsZsrTIKyAGRhYA5rIAWIGQAfGSmAL4A3FYq4TFk8RxcvGRaOmQAQkbcphbxNvaqrE4AhCYAdACSNABy+HgA8lAAogC2AA6yAJ4GFtG2qnEDKonqKWm6AMLZuZbDbvF2xcnGVbX1TW2dPRZkAPxkQoR4NORBslA4MP12Q3ajyZraugAiM2ZzdgWFjs4lALxkXAEMgAMlBZnKXl8ATIALCN0GzHCQA===).

**Comment by stephentoub** [comment]
> In [this](https://godbolt.org/z/5nfdcqh9E) case [the number of returns matches](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+BiAdgVwDZ4CGweMABACYCWAzsaQLABQzAAgMxmsCMSZOWOgDNyrAExkAKjBoAXZgG9mZFV048+wCBDxkAggAoeABjI0AlMtVKmqu1wDsZsrTIKyAGRhYA5rIAWIGQAfGSmAL4A3FYq4TFk8RxcvGRaOmQAQkbcphbxNvaqrE4AhCYAdACSNABy+HgA8lAAogC2AA6yAJ4GFtG2qnEDKonqKWm6AMLZuZbDbvF2xcnGVbX1TW2dPRZkAPxkQoR4NORBslA4MP12Q3ajyZraugAiM2ZzdgWFjs4lALxkXAEMgAMlBZnKXl8ATIALCN0GzHCQA===).

Yes, but again, where in this PR is there code that does that?

**Pull request merged**